src/paginationCalls/pagedApiCaller.ts

```diff
@@ -58,97 +151,131 @@ export class PagedApiCaller implements APICaller {
     this.pageDescriptor = pageDescriptor;
   }
 
-  private createActualCallback(
+  /**
+   * This function translates between regular gRPC calls (that accepts a request and returns a response,
+   * and does not know anything about pages and page tokens) and the users' callback (that expects
+   * to see resources from one page, a request to get the next page, and the raw response from the server).
+   *
+   * It generates a function that can be passed as a callback function to a gRPC call, will understand
+   * pagination-specific fields in the response, and call the users' callback after having those fields
+   * parsed.
+   *
+   * @param request Request object. It needs to be passed to all subsequent next page requests
+   * (the main content of the request object stays unchanged, only the next page token changes)
+   * @param callback The user's callback that expects the page content, next page request, and raw response.
+   */
+  private generateParseResponseCallback(
     request: NextPageRequestType,
     callback: APICallback
   ): APICallback {
-    const self = this;
-    return function fetchNextPageToken(
-      err: Error | null,
-      response: NextPageRequestType | undefined
-    ) {
+    const resourceFieldName = this.pageDescriptor.resourceField;
+    const responsePageTokenFieldName = this.pageDescriptor
+      .responsePageTokenField;
+    const requestPageTokenFieldName = this.pageDescriptor.requestPageTokenField;
+    return (err: Error | null, response: NextPageRequestType | undefined) => {
       if (err) {
         callback(err);
         return;
       }
+      if (!request) {
+        callback(
+          new GoogleError('Undefined request in pagination method callback.')
+        );
+        return;
+      }
       if (!response) {
         callback(
           new GoogleError('Undefined response in pagination method callback.')
         );
         return;
       }
-      const resources = response[self.pageDescriptor.resourceField];
-      const pageToken = response[self.pageDescriptor.responsePageTokenField];
+      const resources = response[resourceFieldName];
+      const pageToken = response[responsePageTokenFieldName];
+      let nextPageRequest = null;
       if (pageToken) {
-        request![self.pageDescriptor.requestPageTokenField] = pageToken;
-        callback(err, resources, request, response);
-      } else {
-        callback(err, resources, null, response);
+        nextPageRequest = Object.assign({}, request);
+        nextPageRequest[requestPageTokenFieldName] = pageToken;
       }
+      callback(err, resources, nextPageRequest, response);
```

Comment:
Here the callback function refers to the private callback in ResourceCollector, is that correct? all the logic is moved to ResourceCollector class.

Reply:
No, here it's the function parameter. It's the callback that was passed.